### ISSUE TITLE
nginx optimisations

### ISSUE THREAD

**Issue by bengarrett** [demozoo/demozoo#382]
Hi, I have been browsing the site using browser Developer Tools. I've noticed some Nginx optimisations that you could quickly implement through simple config file edits.

### gzip compression

demozoo.org has gzip compression enabled for HTML files, but other text transfers such as JS and CSS are sent uncompressed.

Here is an example of a JS file Demozoo serves vs. a transfer over gzip, it's nearly a quarter of the size.

#### uncompressed
![Capture1](https://user-images.githubusercontent.com/513842/62743088-71404e80-ba84-11e9-8954-f4e903683e9b.png)

#### compressed
![Capture2](https://user-images.githubusercontent.com/513842/62743153-ad73af00-ba84-11e9-9755-aaf91b2ca08d.png)

You could add the following to the `http {}` context (usually in `nginx.conf`) to enable compression on text file transfers, even over proxies.

```nginx
# gzip compression
    gzip on;
    gzip_comp_level 5;
    gzip_min_length 256;
    gzip_proxied any;
    gzip_vary on;
    gzip_types text/css
        application/javascript
        image/svg+xml
        text/plain
        text/xml;
```

#### HTTP/2

HTTP2 would require an nginx upgrade to v1.9.5+ but would allow parallel file transfers. It would be useful for Demozoo that sometimes relies on large batches of network requests for thumbnails etc. HTTP2 only works on HTTPS connections and is very easy to enable.

Live demo: https://http2.akamai.com/demo

```nginx
server {
    listen 443 ssl http2;
    server_name demozoo.org;
   ...
}
```

#### Security

By default, nginx displays too much information with its response headers.
![Capture3](https://user-images.githubusercontent.com/513842/62743893-92566e80-ba87-11e9-8d5d-6d88d6ab4e17.png)
This information makes it easy to guess at what distro version you're running, ie LTS 14.04 or earlier which is now unsupported. If there were known vulnerabilities in this distro or nginx version, a bot might go crawling for servers returning this information.

You can remove this by adding `server_tokens off;` to the `http` contexts in the config.
```nginx
http {
   server_tokens off;
   ...
}
```

#### HTTP Strict Transport Security
Demozoo redirects http requests to https. This best practise header, automatically tells modern browsers to always use HTTPS even without redirects.

```nginx
server {
   ...
   add_header Strict-Transport-Security "max-age=63072000; includeSubDomains" always;
}
```

#### X-Content-Type-Options
Just a precaution header "that tells it not to load scripts and stylesheets unless the server indicates the correct MIME type." 
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options

```nginx
server {
   ...
   add_header X-Content-Type-Options "nosniff" always;
}

**Comment by nswaldman** [comment]
These are perfectly reasonable and actionable points. Would definitely recommend we dive into this.

**Comment by sagamusix** [comment]
Just a note on Strict Transport Security - I'm not sure how much your wording is supposed to imply so I will make it explicit: You still need the redirects because the header is *ignored* on HTTP connections. STS just avoids HTTP connections once you have visited the site via HTTPS.

**Comment by bengarrett** [comment]
> Just a note on Strict Transport Security - I'm not sure how much your wording is supposed to imply so I will make it explicit: You still need the redirects because the header is _ignored_ on HTTP connections. STS just avoids HTTP connections once you have visited the site via HTTPS.

Oh absolutely, you should have both the redirects and HSTS. You could even add DZ to [Google's HSTS proload list](https://hstspreload.org/).

> If a website accepts a connection through HTTP and redirects to HTTPS, visitors may initially communicate with the non-encrypted version of the site before being redirected, if, for example, the visitor types http://www.foo.com/ or even just foo.com. This creates an opportunity for a man-in-the-middle attack. The redirect could be exploited to direct visitors to a malicious site instead of the secure version of the original site.

> The HTTP Strict Transport Security header informs the browser that it should never load a site using HTTP and should automatically convert all attempts to access the site using HTTP to HTTPS requests instead.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security

**Comment by gasman** [comment]
Now all implemented in 350d50c2bda07f77964b2ba2f687ddf4164dbde0 + parents. Thanks for the really valuable advice @bengarrett!
